12517 move glossary to bundle

mattamon opened the pull request:
## Changes in this pull request  
Resolves #12517

## Additional info  
- Moving glossary functionality into GlossaryBundle
- Able to install and uninstall via Installer. Permissions are in the Bundle
- User Keybindings are now configurable and can be added from anywhere via event
- Menu Items are now added via Eventlistener pimcore.events.preMenuBuild.
- Menus do now have a priority to figure out the sorting. Standard pimcore menus priority increases by 10 to leave enough space for additonal menu items.

mattamon committed:
Extract glossary from core

mattamon committed:
Remove "use" of wrong installer

mattamon committed:
Adding hook for keybinding registration
Adding migration for settings store
Moving DependencyInjection Configuration back to Core

mattamon committed:
Remove permissions from installer

mattamon committed:
Move translations to bundle

mattamon committed:
Move GlossaryTest back to tests

mattamon committed:
Making user keybindings configurable via yaml

mattamon commented:
What should we do with the documentation?
https://pimcore.com/docs/pimcore/11.0/Development_Documentation/Tools_and_Features/Glossary.html
https://github.com/pimcore/pimcore/blob/11.x/doc/Development_Documentation/18_Tools_and_Features/21_Glossary.md

Is there a way to we can sync this in the documentation after we move it?

brusch commented:
@mattamon since the glossary bundle will stay in this repo, we keep the docs for this in place.
For bundles that will be separated out into their own repo, we move the docs as well, just as we did with other bundles (DataHub, ...)

mattamon committed:
Reorder loading files in jsPaths and extract permissions to constant

mattamon committed:
Add GlossaryBundle to ci/files/config/bundles.php

dvesh3 commented on bundles/GlossaryBundle/public/css/glossary.css:
```diff
@@ -0,0 +1,7 @@
+.pimcore_icon_glossary {
```
filename `icons.css`?

dvesh3 commented on bundles/GlossaryBundle/public/js/glossary.js:
```diff
@@ -11,8 +11,8 @@
 
-pimcore.registerNS("pimcore.settings.glossary");
-pimcore.settings.glossary = Class.create({
+pimcore.registerNS("pimcore.glossary.glossary");
```
can we still use `settings`?

mattamon replied:
Yes we can still use the settings. Just wanted to make it a bit clearer that it is now under glossary instead of settings.

dvesh3 commented on bundles/GlossaryBundle/src/Controller/SettingsController.php:
```diff
@@ -0,0 +1,145 @@
+use Symfony\Component\Routing\Annotation\Route;
+
+/**
+ * @Route("/settings")
```
we are using prefix in the routing.yaml, so the complete route will be `/admin/glossary/settings/glossary` and route name changed to `pimcore_settings_glossary`. Either we keep using the admin prefix in the name or remove it from the actual route prefix. wdyt?

mattamon replied:
I think we cannot remove the admin prefix from the routing, because then we cannot use the AdminController anymore because of the checks. We can simply add the admin back into the name, but I guess we should make it  ```pimcore_admin_glossary_settings_glossary``` then?

dvesh3 replied:
Ok, I just scanned `Ecommercebundle` and there we use the admin prefix for the route and not in the name so keeping it consistent then we can go without admin prefix in the name. could you please revert it? sorry for rework 😞

dvesh3 commented on bundles/GlossaryBundle/src/DependencyInjection/GlossaryExtension.php:
```diff
@@ -0,0 +1,47 @@
+        $loader->load('services.yaml');
+    }
+
+    public function prepend(ContainerBuilder $container): void
```
If this is not implemented, then we can kick `PrependExtensionInterface`

dvesh3 commented on bundles/GlossaryBundle/src/Migrations/Version20221215071650.php:
```diff
@@ -0,0 +1,37 @@
+
+    public function up(Schema $schema): void
+    {
+        $tableExists = Db::get()->fetchOne('SHOW TABLES LIKE "glossary"');
+
+        $installed = !empty($tableExists);
+
+        if ($installed) {
```
```suggestion
        if ($schema->hasTable('glossary')) {
```

dvesh3 replied:
we also need a warning here that the bundle should be enabled manually in `config/bundles.php`.

mattamon replied:
If the glossary table exists:
![image](https://user-images.githubusercontent.com/38959016/209081810-4d3b0eaa-371b-402b-bb71-d1f919a221b3.png)

dvesh3 replied:
looks good.. can we also add the bundle name to the warning?

mattamon replied:
Should we add the whole name ```Pimcore\Bundle\GlossaryBundle\GlossaryBundle```  or just ```GlossaryBundle```?

mattamon replied:
We can do it like this
```php
  $this->warnIf(
          $schema->hasTable('glossary'),
          sprintf('Please make sure to enable the %s manually in config/bundles.php', GlossaryBundle::class)
  );
```

mattamon committed:
Adding feedback changes
Adding warning to enable bundle manually
Removing unnecessary interfaces
Changing glossary.css to icons.css
Changing route names

mattamon committed:
Adding feedback changes
Reverting routes
Adding bundle name to migration warning

mattamon committed:
Adding sorted menu

mattamon committed:
Removing unecessary commas

dvesh3 commented on bundles/GlossaryBundle/src/Migrations/Version20221215071650.php:
```diff
@@ -0,0 +1,34 @@
+/**
+ * Checking if glossary tables already exist
+ */
+final class Version20221215071650 extends AbstractMigration
```
I guess this needs to be moved to Core migrations otherwise we get into chicken-egg problem as the bundle needs to be installed to run the migration and the migration needs to be executed to install the bundle.

dvesh3 reviewed (changes requested):
we should update docs to notify the bundle should be enabled and installed to use the feature https://github.com/pimcore/pimcore/blob/bd8b764859c2a28be11280aab44d20f761c4cb72/doc/Development_Documentation/18_Tools_and_Features/21_Glossary.md#L2-L1

dvesh3 commented on tests/Unit/Document/Glossary/GlossaryTest.php:
```diff
@@ -31,6 +33,9 @@ class GlossaryTest extends TestCase
      */
     protected function setUp(): void
     {
+        if(!GlossaryBundle::isInstalled()) {
```
Glossary tests are [skipped](https://github.com/pimcore/pimcore/actions/runs/3730712672/jobs/6328082354#step:11:3218)at the moment. The bundle needs to be installed before running tests like we do it for ecommerce bundle here https://github.com/pimcore/pimcore/actions/runs/3730712672/jobs/6328082354#step:11:3218

mattamon replied:
Is it enough to write a Helper like this one?
https://github.com/pimcore/pimcore/blob/11.x/tests/Support/Helper/Ecommerce.php

dvesh3 replied:
we already have a helper for Unit suite https://github.com/pimcore/pimcore/blob/56d742deb8e61b33d34040714458460fa774f7fb/tests/Support/Helper/Unit.php#L23

dvesh3 replied:
maybe we don't need the skip check anymore since the bundle is always installed.

mattamon committed:
Move migration to core/migrations

mattamon committed:
Update Documentation, to enable and install the bundle

mattamon committed:
Installing GlossaryBundle for testing

mattamon committed:
Refactoring installation to own method.
Removing glossary installed check

mattamon committed:
Merge branch '11.x' into 12517-move-glossary-to-bundle

mattamon commented:
Updating docs with new examples -> https://pimcore.com/docs/pimcore/11.0/Development_Documentation/Best_Practice/Open_By-External_Id.html

mattamon committed:
Moving translation back into core
Removing if condition from buildmenu

mattamon committed:
Rename bundle from GlossaryBundle to PimcoreGlossaryBundle

mattamon committed:
Adding Upgrade notes

dvesh3 commented on doc/Development_Documentation/23_Installation_and_Upgrade/09_Upgrade_Notes/README.md:
```diff
@@ -55,7 +55,10 @@ Please make sure to set your preferred storage location ***before*** migration.
-- [Glossary] `pimcoreglossary()` tag has been removed, please use the `pimcore_glossary` Twig filter.
+- [Glossary] 
+  - `pimcoreglossary()` tag has been removed, please use the `pimcore_glossary` Twig filter.
+  -  Glossary was moved into PimcoreGlossaryBundle
```
```suggestion
  -  Glossary has been moved into PimcoreGlossaryBundle
```

dvesh3 committed:
Update doc/Development_Documentation/23_Installation_and_Upgrade/09_Upgrade_Notes/README.md

mattamon committed:
Adding extras menu by default and do not create if no items

brusch commented on bundles/AdminBundle/config/pimcore/default.yaml:
```diff
@@ -19,6 +19,135 @@ pimcore_admin:
+                key: 'O'
+                ctrl: true
+                shift: true
+            openDocument:
```
```suggestion
            open_document:
```
I think we should use snake-case here for all the keys, as this is Symfony bast practice and we should be consistent here.

mattamon replied:
Yes this seems to be the right way. I Just did it to match the method names. Maybe we should add the method then as its  own property? I don't want to like fiddle around with all the keys to make proper function calls automatically. I would then set the key and the action as required
```yaml
            open_document:
                key: 'D'
                action: openDocument
                ctrl: true
                shift: true
```

Something like this?

dvesh3 replied:
fixed it with adding `action` property

brusch commented on bundles/AdminBundle/public/js/pimcore/helpers.js:
```diff
@@ -3241,3 +3236,37 @@ pimcore.helpers.deleteConfirm = function (title, name, deleteCallback) {
+ * Building menu with priority
+ * @param items
+ */
+pimcore.helpers.buildMenu = function(items) {
```
Do we really still need that, as we're doing the ordering in `pimcore.menu.menu` anyway?

mattamon replied:
Yes the buildMenu runs through all submenus since every submenu needs to be a pimcore.menu.menu, as well. Otherwise it  won't get sorted.

brusch commented on bundles/AdminBundle/public/js/pimcore/helpers.js:
```diff
@@ -3241,3 +3236,37 @@ pimcore.helpers.deleteConfirm = function (title, name, deleteCallback) {
+ * @returns {pimcore.menu.menu}
+ */
+pimcore.helpers.createMenu = function(menu) {
+    return Ext.create('pimcore.menu.menu', menu);
```
I think we don't need a helper for this, let's just use `Ext.create('pimcore.menu.menu', menu)` directly 😊

brusch commented on bundles/AdminBundle/public/js/pimcore/layout/menu.js:
```diff
@@ -0,0 +1,43 @@
+ */
+
+/**
+ * Adding a priority sorting function for menus
```
```suggestion
 * @internal
 * Adding a priority sorting function for menus
```

brusch commented on bundles/GlossaryBundle/config/pimcore/default.yaml:
```diff
@@ -0,0 +1,15 @@
+# Pimcore
+pimcore:
+    glossary:
```
```suggestion
pimcore_glossary:
```
To be consistent, we probably should use a dedicated root key here, right? 
The config definition needs to be moved to the bundle as well. 
An we need to consider this in the upgrade notes.

brusch commented on bundles/GlossaryBundle/public/js/glossary.js:
```diff
@@ -57,7 +57,7 @@ pimcore.settings.glossary = Class.create({
         var itemsPerPage = pimcore.helpers.grid.getDefaultPageSize();
         this.store = pimcore.helpers.grid.buildDefaultStore(
-            Routing.generate('pimcore_admin_settings_glossary'),
+            Routing.generate('pimcore_glossary_settings_glossary'),
```
maybe we should move that to the `pimcore.glossary.*` namespace as well, to be consistent.

mattamon replied:
Well strange, this is already in the glossary namespace in my file. :) 
Otherwise the routing would not work because I changed the route in the Settingscontroller too.

mattamon commented:
Added a section in the Upgrade Notes for all bundle extractions

mattamon committed:
Adding method to key_bindings
Adding configuration for pimcore_glossary
Removing helper function

mattamon committed:
Renaming method to action
Removing class dependency in migration

mattamon committed:
Adding defaultkeybinding definition
